walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,120 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    
+    Collection<TimeRange> answer = new ArrayList<TimeRange>();
+
+    Collection<String> requestAttendees = new ArrayList<String>(request.getAttendees());
+    requestAttendees.addAll(request.getOptionalAttendees());
+    long requestDuration = request.getDuration();
+    
+    List<TimeRange> relevantEventTimes = filterEvents(events, requestAttendees);
+
+    System.out.println(relevantEventTimes);
+    findPossibleTimes(answer, relevantEventTimes, requestDuration);
+    
+    //If accommodating optional attendees isn't possible, only consider mandatory attendees
+    if (answer.size() == 0 && request.getAttendees().size() > 0) {
+      requestAttendees = request.getAttendees();  
+      relevantEventTimes = filterEvents(events, requestAttendees);
+      findPossibleTimes(answer, relevantEventTimes, requestDuration);
+    }
+    
+    return answer;
+  }
+ 
+  /**
+   * Helper function that updates a list of available times for meeting, given
```

Comment:
Per style guide, the document should be a complete sentence by itself.

Reply:
yes, fixed